### PR TITLE
Add invoked functions to query completed event

### PR DESCRIPTION
**Pull request by zacw7**
Add the names of the functions invoked by the query to the completed event when session property `log_invoked_function_names_enabled` or system config `log-invoked-function-names-enabled` is enabled.

Check `com.facebook.presto.hive.TestHiveIntegrationSmokeTest#testInvokedFunctionNamesLog` for details.

```
== RELEASE NOTES ==

General Changes
* Add the names of the functions invoked by the query to the completed event.
This can be enabled with the ``log_invoked_function_names_enabled`` session property or the ``log-invoked-function-names-enabled`` configuration property.
```

### CODE REVIEW

File: presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 1]
```diff
@@ -900,6 +901,11 @@ public Optional<QuerySpecification> getCurrentQuerySpecification()
         return currentQuerySpecification;
     }
 
+    public List<String> getInvokedFunctionNames()
+    {
+        return ImmutableList.copyOf(functionHandles.values().stream().map(functionHandle -> functionHandle.getName()).collect(toSet()));
```
Comment (NikhilCollooru) [inline presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:906]: you can use` .map(FunctionHandle::getCatalogSchemaName)`

Comment (NikhilCollooru) [reply presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I meant 
`return functionHandles.values().stream().map(FunctionHandle::getName).collect(toImmutableList());`
why do we need another copy again ?

**Comment by NikhilCollooru** [comment]
The changes looks good. 
Do we want to limit the size of the the list of functions. like can there be a query with 1000 function names? ..if so it would blow up the size of the event object. wondering if it makes sense to limit the size.

File: presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 1 above)
Comment (NikhilCollooru) [reply presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Ahh okay I see it. we are using set to de-duplicate and then making a list out of it

**Comment by zacw7** [comment]
> The changes looks good. Do we want to limit the size of the the list of functions. like can there be a query with 1000 function names? ..if so it would blow up the size of the event object. wondering if it makes sense to limit the size. like may be we log only 20 function names or so ?

Thanks for reviewing the PR! I think it should fine.

The longest function I found so far is an internal UDF which has 44 characters, let's add the prefix `presto.default` and make it 59 characters long. Assuming 1000 such functions are used in one query, it still only takes 59K characters in total while another field `runtime_stats` can be as long as 790K (e.g. 20230202_133039_02539_hw725). I tried "SHOW FUNCTIONS" on a verifier and there were only 998 functions and a lot of them actually share the same name (with different parameters) and hence would be counted only once. So I think it's very unlikely for the size to be larger than 30K.

That said, we can always have the feature tested on testing clusters first, populate some logs with this new added column and check if the size needs to be limited.

**Comment by NikhilCollooru** [comment]
> > The changes looks good. Do we want to limit the size of the the list of functions. like can there be a query with 1000 function names? ..if so it would blow up the size of the event object. wondering if it makes sense to limit the size. like may be we log only 20 function names or so ?
> 
> Thanks for reviewing the PR! I think it should fine.
> 
> The longest function I found so far is an internal UDF which has 44 characters, let's add the prefix `presto.default` and make it 59 characters long. Assuming 1000 such functions are used in one query, it still only takes 59K characters in total while another field `runtime_stats` can be as long as 790K (e.g. 20230202_133039_02539_hw725). I tried `SHOW FUNCTIONS` on a verifier and there were only 998 functions and a lot of them actually share the same name (with different parameters) and hence would be counted only once. So I think it's very unlikely for the size to be larger than even 20K.
> 
> That said, we can always have the feature tested on testing clusters first, populate some logs with this new added column and check if the size needs to be limited.

Got it. sounds good.

**Review by NikhilCollooru — APPROVED** [review]

**Pull request merged**